### ISSUE TITLE
ASYNCIFY=1 unnecessarily generates many temporary locals

### ISSUE THREAD

**Issue by kichikuou** [emscripten-core/emscripten#22196]
`-s ASYNCIFY=1` in Emscripten 3.1.60 and later generates more local variables than before.

`test.c`:
```c
#include <emscripten.h>

EM_ASYNC_JS(int, some_async_func, (int x), {
    return x;
});

int main() {
    int a = 0;
    a = some_async_func(a);
    a = some_async_func(a);
    a = some_async_func(a);
    a = some_async_func(a);
    a = some_async_func(a);
    return a;
}
```
Command line to build:
```
emcc -O3 -sASYNCIFY=1 -o test.js test.c
```

`main()` compiled with emscripten 3.1.60 has two params and six locals, where locals 3 through 7 are used only once each to temporarily hold the result of each `some_async_func()`.

<details>
<summary>`main()` compiled with emscripten 3.1.60</summary>

```wast
  (func (;10;) (type 3) (param i32 i32) (result i32)
    (local i32 i32 i32 i32 i32 i32)
    global.get 1
    i32.const 2
    i32.eq
    if  ;; label = @1
      global.get 2
      global.get 2
      i32.load
      i32.const 4
      i32.sub
      i32.store
      global.get 2
      i32.load
      i32.load
      local.set 0
    end
    block (result i32)  ;; label = @1
      global.get 1
      i32.const 2
      i32.eq
      if  ;; label = @2
        global.get 2
        global.get 2
        i32.load
        i32.const 4
        i32.sub
        i32.store
        global.get 2
        i32.load
        i32.load
        local.set 2
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.eqz
      i32.or
      if  ;; label = @2
        i32.const 0
        call 0
        local.set 3
        i32.const 0
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 3
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 1
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 4
        i32.const 1
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 4
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 2
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 5
        i32.const 2
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 5
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 3
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 6
        i32.const 3
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 6
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 4
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 7
        i32.const 4
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 7
        local.set 0
      end
      global.get 1
      i32.eqz
      if  ;; label = @2
        local.get 0
        return
      end
      unreachable
    end
    local.set 1
    global.get 2
    i32.load
    local.get 1
    i32.store
    global.get 2
    global.get 2
    i32.load
    i32.const 4
    i32.add
    i32.store
    global.get 2
    i32.load
    local.get 0
    i32.store
    global.get 2
    global.get 2
    i32.load
    i32.const 4
    i32.add
    i32.store
    i32.const 0)
```
</details>

When built with Emscripten 3.1.59, these temp locals are coalesced and `main()` has only one local.

<details>
<summary>`main()` compiled with emscripten 3.1.59</summary>

```wast
  (func (;10;) (type 3) (param i32 i32) (result i32)
    (local i32)
    global.get 1
    i32.const 2
    i32.eq
    if  ;; label = @1
      global.get 2
      global.get 2
      i32.load
      i32.const 4
      i32.sub
      i32.store
      global.get 2
      i32.load
      i32.load
      local.set 0
    end
    block (result i32)  ;; label = @1
      global.get 1
      i32.const 2
      i32.eq
      if  ;; label = @2
        global.get 2
        global.get 2
        i32.load
        i32.const 4
        i32.sub
        i32.store
        global.get 2
        i32.load
        i32.load
        local.set 2
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.eqz
      i32.or
      if  ;; label = @2
        i32.const 0
        call 0
        local.set 1
        i32.const 0
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 1
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 1
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 1
        i32.const 1
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 1
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 2
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 1
        i32.const 2
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 1
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 3
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 1
        i32.const 3
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 1
        local.set 0
      end
      global.get 1
      i32.eqz
      local.get 2
      i32.const 4
      i32.eq
      i32.or
      if  ;; label = @2
        local.get 0
        call 0
        local.set 1
        i32.const 4
        global.get 1
        i32.const 1
        i32.eq
        br_if 1 (;@1;)
        drop
        local.get 1
        local.set 0
      end
      global.get 1
      i32.eqz
      if  ;; label = @2
        local.get 0
        return
      end
      unreachable
    end
    local.set 1
    global.get 2
    i32.load
    local.get 1
    i32.store
    global.get 2
    global.get 2
    i32.load
    i32.const 4
    i32.add
    i32.store
    global.get 2
    i32.load
    local.get 0
    i32.store
    global.get 2
    global.get 2
    i32.load
    i32.const 4
    i32.add
    i32.store
    i32.const 0)
```
</details>

I noticed this because a function that previously had 34 local variables now has 444 locals when compiled with the latest Emscripten (3.1.62).

**Comment by kichikuou** [comment]
I have bisected Binaryen. (Emscripten 3.1.59 + local Binaryen build)

This seems to have regressed in https://github.com/WebAssembly/binaryen/commit/7b2e0190213487b5d2505fe86aa9bbbd30e80fcc.

**Comment by sbc100** [comment]
@tlively @kripken

**Comment by kripken** [comment]
Thanks for bisecting this @kichikuou I investigated this and found that what happens here is a weird interaction with StackIR. After that commit, we run StackIR in more situations, which is normally good. However, emscripten then runs some additional transformations on the wasm, like it does metadce and then runs a pass to remove parts of the wasm that we don't need. It re-reads the wasm and writes it to do that. The problem is that more StackIR optimizations lead to more "stacky" code, and that code leads to additional locals being added during the read. Those additional locals would be optimized away easily, but when we do things like remove code after metadce we don't re-run the entire optimization pipeline. That is, the pipeline is

* `-O3`, write a wasm
* Read the wasm, remove code based on metadce results, write a second wasm

If we also did `-O3` in the second line there, then this problem would not happen.

As this is pretty rare and has a small effect on code size I'm not sure if we should make such a change, though, since running all of `-O3` takes time. But, you say that you are also seeing a case with 34 locals that goes to 444, which sounds a lot more worrying than the extra 5 in the testcase. Just to check, that was 444 and not 44? And do you see an impact on code size or speed?

**Comment by kichikuou** [comment]
Yes, it was 444, not 44. It was a [function](https://github.com/kichikuou/xsystem4/blob/cff2271378b3b5c82d2389f7bde12cdd41f1d9ed/src/vm.c#L749) with a huge switch statement that executes virtual machine instructions.

I found this problem while investigating a ["Maximum call stack size exceeded" error in iOS Safari](https://github.com/kichikuou/xsystem4-web/issues/1). It turned out that this local variables issue was not the root cause of the error, but I thought I would report it anyway.

The change in generated code size from 3.1.59 to 3.1.60 was small (less than 1%), although the latter has ~27% more local variables:
```
$ grep '(local' xsystem4.3159.wat |wc
    4657   37651  187860
$ grep '(local' xsystem4.3160.wat |wc
    4657   47949  229052
```

The impact on speed was unnoticeable.

In summary, this issue isn't causing me any real troubles, so far. :)

**Comment by kripken** [comment]
Thanks @kichikuou , that's good to know. It seems worrying enough that I think we should fix it, and I opened https://github.com/emscripten-core/emscripten/pull/22218 for that now, which avoids this issue by running StackIR at the very end.

**Comment by kichikuou** [comment]
I've confirmed that Emscripten 3.1.64 optimizes out those local variables.
Thank you!
